zetasql/src/main/scala/no/nrk/bigquery/ZetaSql.scala

```diff
@@ -190,56 +249,101 @@ object ZetaSql {
-          )
-        case BQField.Type.ARRAY =>
-          TypeFactory.createArrayType(toType(field.subFields.head))
-        case BQField.Type.TIMESTAMP => TypeFactory.createSimpleType(TypeKind.TYPE_TIMESTAMP)
-        case BQField.Type.DATE => TypeFactory.createSimpleType(TypeKind.TYPE_DATE)
-        case BQField.Type.TIME => TypeFactory.createSimpleType(TypeKind.TYPE_TIME)
-        case BQField.Type.DATETIME => TypeFactory.createSimpleType(TypeKind.TYPE_DATETIME)
-        case BQField.Type.GEOGRAPHY => TypeFactory.createSimpleType(TypeKind.TYPE_GEOGRAPHY)
-        case BQField.Type.JSON => TypeFactory.createSimpleType(TypeKind.TYPE_JSON)
-        case BQField.Type.INTERVAL => TypeFactory.createSimpleType(TypeKind.TYPE_INTERVAL)
-      }
-      if (isArray) TypeFactory.createArrayType(elemType) else elemType
+  private def toType(field: BQField): Type = {
+    val isArray = field.mode == BQField.Mode.REPEATED
+
+    val elemType = field.tpe match {
+      case BQField.Type.BOOL => TypeFactory.createSimpleType(TypeKind.TYPE_BOOL)
+      case BQField.Type.INT64 => TypeFactory.createSimpleType(TypeKind.TYPE_INT64)
+      case BQField.Type.FLOAT64 => TypeFactory.createSimpleType(TypeKind.TYPE_FLOAT)
+      case BQField.Type.NUMERIC => TypeFactory.createSimpleType(TypeKind.TYPE_NUMERIC)
+      case BQField.Type.BIGNUMERIC => TypeFactory.createSimpleType(TypeKind.TYPE_BIGNUMERIC)
+      case BQField.Type.STRING => TypeFactory.createSimpleType(TypeKind.TYPE_STRING)
+      case BQField.Type.BYTES => TypeFactory.createSimpleType(TypeKind.TYPE_BYTES)
+      case BQField.Type.STRUCT =>
+        TypeFactory.createStructType(
+          field.subFields
+            .map(sub => new StructType.StructField(sub.name, toType(sub)))
+            .asJavaCollection
+        )
+      case BQField.Type.ARRAY =>
+        TypeFactory.createArrayType(toType(field.subFields.head))
+      case BQField.Type.TIMESTAMP => TypeFactory.createSimpleType(TypeKind.TYPE_TIMESTAMP)
+      case BQField.Type.DATE => TypeFactory.createSimpleType(TypeKind.TYPE_DATE)
+      case BQField.Type.TIME => TypeFactory.createSimpleType(TypeKind.TYPE_TIME)
+      case BQField.Type.DATETIME => TypeFactory.createSimpleType(TypeKind.TYPE_DATETIME)
+      case BQField.Type.GEOGRAPHY => TypeFactory.createSimpleType(TypeKind.TYPE_GEOGRAPHY)
+      case BQField.Type.JSON => TypeFactory.createSimpleType(TypeKind.TYPE_JSON)
+      case BQField.Type.INTERVAL => TypeFactory.createSimpleType(TypeKind.TYPE_INTERVAL)
     }
+    if (isArray) TypeFactory.createArrayType(elemType) else elemType
+  }
+
+  def toTableValuedFunction(atvf: BQAppliedTableValuedFunction[Any]): TVFInfo = {
+    val tableType = new FunctionArgumentType(SignatureArgumentKind.ARG_TYPE_RELATION)
+    val args = new java.util.ArrayList(
+      atvf.params
+        .map { p =>
+          new FunctionArgumentType(
+            p.maybeType
+              .map { p =>
+                def toType2(tpe: BQField.Type): Type = tpe match {
```

Comment:
this and toType(bqField) should be one.
toTypebqField should call the new method, since it has one other dimention in Mode

Reply:
Fixed.